### PR TITLE
add memory hooks for cheat/rumble functionality

### PR DESCRIPTION
**Pull request by RetroSven**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,6 +12,16 @@ src/common/scm_rev.cpp
 .vs/
 .vscode/
 CMakeLists.txt.user*
+src/CMakeCache.txt
```
Comment (j-selby) [inline .gitignore]: These files don't need to be added to the .gitignore file. If they come up in your own builds, you should just add them to your clone's own gitignore.

File: src/citra_libretro/citra_libretro.cpp [hunk 2]
```diff
@@ -513,12 +515,16 @@ bool retro_unserialize(const void* data_, size_t size) {
 }
 
 void* retro_get_memory_data(unsigned id) {
-    (void)id;
+    if ( id == RETRO_MEMORY_SYSTEM_RAM )
+        return Kernel::memory_regions[0].linear_heap_memory->data() ;
```
Comment (j-selby) [inline src/citra_libretro/citra_libretro.cpp]: No space between `data()` and `;`.

File: src/citra_libretro/citra_libretro.cpp [hunk 3]
```diff
@@ -513,12 +515,16 @@ bool retro_unserialize(const void* data_, size_t size) {
...
+
     return NULL;
 }
 
 size_t retro_get_memory_size(unsigned id) {
-    (void)id;
+    if ( id == RETRO_MEMORY_SYSTEM_RAM )
+        return Kernel::memory_regions[0].size ;
```
Comment (j-selby) [inline src/citra_libretro/citra_libretro.cpp]: Ditto.

File: .gitignore (hunk 1 above)
Comment (RetroSven) [reply .gitignore]: Do they cause a problem?  I figured that if they affected me they would potentially affect others.

**Comment by RetroSven** [comment]
What is the process you'd like me to follow for these changes?  Just submit a new PR with the changes?

Direct access to the memory is required in order for the RA cheat/rumble system to function correctly.  It needs to read/write the actual memory contents.

**Comment by j-selby** [comment]
`What is the process you'd like me to follow for these changes? Just submit a new PR with the changes?`

Up to you. Feel free to force push + update this PR - I have no qualms with that.

`Direct access to the memory is required in order for the RA cheat/rumble system to function correctly.`

Well, there is no guarantee that Citra's memory is going to be a linear blob. That model, as was pointed out to me, isn't even going to work with multiprocess support.

**Comment by RetroSven** [comment]
> Well, there is no guarantee that Citra's memory is going to be a linear blob. That model, as was pointed out to me, isn't even going to work with multiprocess support.

I noticed that and compromised on just pulling the "application" section of the fcram since it was divided into three distinct chunks and it seems most likely that's where the interesting bits will be.  It's a "better than nothing" solution for now and will potentially give some entertainment to the end-users.

I have future plans for RA accepting more than a single memory location which won't involve changes to the libretro API.  A more generic approach of implementing getmem/putmem API calls would future-proof it entirely - but that's likely more work than is necessary because for 99% of the emulators in RA, the current method of a single contiguous blob works well.

**Comment by RetroSven** [comment]
ok added commit here - thanks

**Comment by j-selby** [comment]
As long as that obvious caveat is acknowledged, I'll let this go through.

**Pull request merged**